<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Clustering on Network of Samples</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Clustering on Network of Samples
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;conos&rsquo; version 1.4.2</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="basicSeuratProc.html">basicSeuratProc</a></td>
<td>Create and preprocess a Seurat object</td></tr>
<tr><td style="width: 25%;"><a href="bestClusterThresholds.html">bestClusterThresholds</a></td>
<td>Find threshold of cluster detectability</td></tr>
<tr><td style="width: 25%;"><a href="bestClusterTreeThresholds.html">bestClusterTreeThresholds</a></td>
<td>Find threshold of cluster detectability in trees of clusters</td></tr>
<tr><td style="width: 25%;"><a href="buildWijMatrix.html">buildWijMatrix</a></td>
<td>Rescale the weights in an edge matrix to match a given perplexity.</td></tr>
<tr><td style="width: 25%;"><a href="buildWijMatrix.html">buildWijMatrix.CsparseMatrix</a></td>
<td>Rescale the weights in an edge matrix to match a given perplexity.</td></tr>
<tr><td style="width: 25%;"><a href="buildWijMatrix.html">buildWijMatrix.TsparseMatrix</a></td>
<td>Rescale the weights in an edge matrix to match a given perplexity.</td></tr>
<tr><td style="width: 25%;"><a href="Conos.html">Conos</a></td>
<td>Conos R6 class</td></tr>
<tr><td style="width: 25%;"><a href="convertToPagoda2.html">convertToPagoda2</a></td>
<td>Convert Conos object to Pagoda2 object</td></tr>
<tr><td style="width: 25%;"><a href="edgeMat.html">edgeMat</a></td>
<td>Set edge matrix edgeMat with certain values on sample</td></tr>
<tr><td style="width: 25%;"><a href="edgeMat.html">edgeMat-method</a></td>
<td>Set edge matrix edgeMat with certain values on sample</td></tr>
<tr><td style="width: 25%;"><a href="edgeMat.html">edgeMat&lt;-</a></td>
<td>Set edge matrix edgeMat with certain values on sample</td></tr>
<tr><td style="width: 25%;"><a href="edgeMat.html">edgeMat&lt;--method</a></td>
<td>Set edge matrix edgeMat with certain values on sample</td></tr>
<tr><td style="width: 25%;"><a href="estimateWeightEntropyPerCell.html">estimateWeightEntropyPerCell</a></td>
<td>Estimate entropy of edge weights per cell according to the specified factor. Can be used to visualize alignment quality according to this factor.</td></tr>
<tr><td style="width: 25%;"><a href="findSubcommunities.html">findSubcommunities</a></td>
<td>Increase resolution for a specific set of clusters</td></tr>
<tr><td style="width: 25%;"><a href="getBetweenCellTypeCorrectedDE.html">getBetweenCellTypeCorrectedDE</a></td>
<td>Compare two cell types across the entire panel</td></tr>
<tr><td style="width: 25%;"><a href="getBetweenCellTypeDE.html">getBetweenCellTypeDE</a></td>
<td>Compare two cell types across the entire panel</td></tr>
<tr><td style="width: 25%;"><a href="getCellNames.html">getCellNames</a></td>
<td>Access cell names from sample</td></tr>
<tr><td style="width: 25%;"><a href="getCellNames.html">getCellNames-method</a></td>
<td>Access cell names from sample</td></tr>
<tr><td style="width: 25%;"><a href="getClustering.html">getClustering</a></td>
<td>Access clustering from sample</td></tr>
<tr><td style="width: 25%;"><a href="getClustering.html">getClustering-method</a></td>
<td>Access clustering from sample</td></tr>
<tr><td style="width: 25%;"><a href="getCountMatrix.html">getCountMatrix</a></td>
<td>Access count matrix from sample</td></tr>
<tr><td style="width: 25%;"><a href="getCountMatrix.html">getCountMatrix-method</a></td>
<td>Access count matrix from sample</td></tr>
<tr><td style="width: 25%;"><a href="getEmbedding.html">getEmbedding</a></td>
<td>Access embedding from sample</td></tr>
<tr><td style="width: 25%;"><a href="getEmbedding.html">getEmbedding-method</a></td>
<td>Access embedding from sample</td></tr>
<tr><td style="width: 25%;"><a href="getGeneExpression.html">getGeneExpression</a></td>
<td>Access gene expression from sample</td></tr>
<tr><td style="width: 25%;"><a href="getGeneExpression.html">getGeneExpression-method</a></td>
<td>Access gene expression from sample</td></tr>
<tr><td style="width: 25%;"><a href="getGenes.html">getGenes</a></td>
<td>Access genes from sample</td></tr>
<tr><td style="width: 25%;"><a href="getGenes.html">getGenes-method</a></td>
<td>Access genes from sample</td></tr>
<tr><td style="width: 25%;"><a href="getOverdispersedGenes.html">getOverdispersedGenes</a></td>
<td>Access overdispersed genes from sample</td></tr>
<tr><td style="width: 25%;"><a href="getOverdispersedGenes.html">getOverdispersedGenes-method</a></td>
<td>Access overdispersed genes from sample</td></tr>
<tr><td style="width: 25%;"><a href="getPca.html">getPca</a></td>
<td>Access PCA from sample</td></tr>
<tr><td style="width: 25%;"><a href="getPca.html">getPca-method</a></td>
<td>Access PCA from sample</td></tr>
<tr><td style="width: 25%;"><a href="getPerCellTypeDE.html">getPerCellTypeDE</a></td>
<td>Do differential expression for each cell type in a conos object between the specified subsets of apps</td></tr>
<tr><td style="width: 25%;"><a href="getRawCountMatrix.html">getRawCountMatrix</a></td>
<td>Access raw count matrix from sample</td></tr>
<tr><td style="width: 25%;"><a href="getRawCountMatrix.html">getRawCountMatrix-method</a></td>
<td>Access raw count matrix from sample</td></tr>
<tr><td style="width: 25%;"><a href="getSampleNamePerCell.html">getSampleNamePerCell</a></td>
<td>Retrieve sample names per cell</td></tr>
<tr><td style="width: 25%;"><a href="greedyModularityCut.html">greedyModularityCut</a></td>
<td>Performs a greedy top-down selective cut to optmize modularity</td></tr>
<tr><td style="width: 25%;"><a href="p2app4conos.html">p2app4conos</a></td>
<td>Utility function to generate a pagoda2 app from a conos object</td></tr>
<tr><td style="width: 25%;"><a href="plotClusterBarplots.html">plotClusterBarplots</a></td>
<td>Plots barplots per sample of composition of each pagoda2 application based on selected clustering</td></tr>
<tr><td style="width: 25%;"><a href="plotClusterBoxPlotsByAppType.html">plotClusterBoxPlotsByAppType</a></td>
<td>Generate boxplot per cluster of the proportion of cells in each celltype</td></tr>
<tr><td style="width: 25%;"><a href="plotComponentVariance.html">plotComponentVariance</a></td>
<td>Plot fraction of variance explained by the successive reduced space components (PCA, CPCA)</td></tr>
<tr><td style="width: 25%;"><a href="plotDEheatmap.html">plotDEheatmap</a></td>
<td>Plot a heatmap of differential genes</td></tr>
<tr><td style="width: 25%;"><a href="projectKNNs.html">projectKNNs</a></td>
<td>Project a distance matrix into a lower-dimensional space.</td></tr>
<tr><td style="width: 25%;"><a href="rawMatricesWithCommonGenes.html">rawMatricesWithCommonGenes</a></td>
<td>Get raw matrices with common genes</td></tr>
<tr><td style="width: 25%;"><a href="saveConosForScanPy.html">saveConosForScanPy</a></td>
<td>Save Conos object on disk to read it from ScanPy</td></tr>
<tr><td style="width: 25%;"><a href="saveDEasCSV.html">saveDEasCSV</a></td>
<td>Save differential expression as table in *csv format</td></tr>
<tr><td style="width: 25%;"><a href="saveDEasJSON.html">saveDEasJSON</a></td>
<td>Save differential expression results as JSON</td></tr>
<tr><td style="width: 25%;"><a href="scanKModularity.html">scanKModularity</a></td>
<td>Scan joint graph modularity for a range of k (or k.self) values Builds graph with different values of k (or k.self if scan.k.self=TRUE), evaluating modularity of the resulting multilevel clustering NOTE: will run evaluations in parallel using con$n.cores (temporarily setting con$n.cores to 1 in the process)</td></tr>
<tr><td style="width: 25%;"><a href="sgdBatches.html">sgdBatches</a></td>
<td>Calculate the default number of batches for a given number of vertices and edges. The formula used is the one used by the 'largeVis' reference implementation.  This is substantially less than the recommendation E * 10000 in the original paper.</td></tr>
<tr><td style="width: 25%;"><a href="small_panel.preprocessed.html">small_panel.preprocessed</a></td>
<td>Small pre-processed data from Pagoda2, two samples, each dimension (1000, 100)</td></tr>
<tr><td style="width: 25%;"><a href="stableTreeClusters.html">stableTreeClusters</a></td>
<td>Determine number of detectable clusters given a reference walktrap and a bunch of permuted walktraps</td></tr>
</table>
</body></html>
